### ISSUE TITLE
Cannot publish unborn HEAD/Commit failed exit code 128 'cannot lock ref 'HEAD''.

### ISSUE THREAD

**Issue by cfandel** [desktop/desktop#7581]
## Description

Although several existing issues seem similar, none actually address my problem.

I am working in an existing repository that I have been committing and pushing changes to successfully for several months. I am the only person contributing. Overnight, I stopped being able to commit, and get the following error message:

```
'Commit failed - exit code 128 received, with output: 'fatal: cannot lock ref 'HEAD': unable to resolve reference 'refs/heads/master': reference broken'
```

My current branch should be 'master' but has changed itself to '(unknown)'. When I try to change it back, I get the following error message:

```
'Unable to switch branches as there are working directory changes which would be overwritten. Please commit or stash your changes.'
```

I have local changes I want to keep, because I made changes before the error occurred and haven't been able to commit them. 

Screenshots:
![image](https://user-images.githubusercontent.com/31896726/58026330-860ce600-7b16-11e9-8d08-c035214ca630.png)
![image](https://user-images.githubusercontent.com/31896726/58026355-9329d500-7b16-11e9-8b3d-597526f50477.png)

## Version

* GitHub Desktop:

* Operating system:
Windows v10.0.17134.765
GitHub Desktop v1.6.6

## Steps to Reproduce

Can't reproduce as this is only happening on one repository and I don't know what caused it.

**Comment by steveward** [comment]
Thanks for the report @cfandel. It looks like there may be some local repository corruption (hence the `broken reference` error). Could you go to the menu and select `Repository` > `Open in <command_line>` and run the command `git fsck`? Let me know the full output of that command -- it should give some more information about any issues with your local repository.

**Comment by cfandel** [comment]
Thank for the quick response!
Here is a screenshot of the results of `git fsck` in Git Bash (gives same result as in Command Prompt).
![image](https://user-images.githubusercontent.com/31896726/58072852-4e974b80-7ba1-11e9-81a1-8f4dc416f351.png)

**Comment by steveward** [comment]
@cfandel thanks for sharing that output.

Since you are the only person working with this repository I would recommend [following this StackOverflow answer](https://stackoverflow.com/a/46604551) which shares the specific commands you can run to move your corrupt `.git` directory and replace it with a fresh copy from the GitHub. These steps will leave your uncommitted changes in place. I would recommend making a backup of your existing repository in case anything goes wrong during that process. 

If you run into any issues please let me know.

**Comment by cfandel** [comment]
OK thanks - just to clarify though, the version I want to keep is the local version, the one stored on my desktop computer. I really really do not want to overwrite my local files with the version stored on GitHub. 
Does the solution you linked to (the answer by Zoey Hewll) preserve the local version or the version stored on GitHub? And to clarify terminology, the version stored on GitHub is called the remote version right?

Would it be easier for me to just completely delete the repository on GitHub, and create a new one from scratch by uploading my local files?

**Comment by cfandel** [comment]
Also, side question: what causes local repositories to be corrupted and how do I avoid it in the future?

**Comment by steveward** [comment]
>Does the solution you linked to (the answer by Zoey Hewll) preserve the local version or the version stored on GitHub? And to clarify terminology, the version stored on GitHub is called the remote version right?

The solution I linked keeps your local files intact -- it just removes the corrupt local `.git` directory and replaces it with a fresh copy from the remote repository on GitHub.

>And to clarify terminology, the version stored on GitHub is called the remote version right?

That's correct. The term `remotes` refers to copies of your repository hosted in different locations (on the internet, on a local server, etc.).

>Would it be easier for me to just completely delete the repository on GitHub, and create a new one from scratch by uploading my local files?

I wouldn't recommend doing that since the version on GitHub is not corrupt (GitHub does not allow you to push corrupt objects up to your remote repository). If you have a more recent backup of your repository somewhere else you could use that for recovery as well.

>Also, side question: what causes local repositories to be corrupted and how do I avoid it in the future?

Common causes that I have seen are sudden power failures or third-party file syncing services like Dropbox being used as the storage directory for repositories.

I hope that is helpful!

**Comment by tierninho** [comment]
Closing this issue up for now as it hopefully has been resolved. @cfandel should you have any more details or questions, please reach out and we will be happy to help. Thank you.

**Comment by cfandel** [comment]
Hi - sorry for the delay - I was on vacation last week.
I actually do still have issues with this.
First - when running the commands in the post you linked to:
```
mv -v .git .git_old &&            # remove old git
git init &&                       # initialise new repo
git remote add origin "${url}" && # link to old repo
git fetch &&                      # get old history
git reset origin/master --mixed   # force update to old history
```

I get the following error message:
```
renamed '.git'-> '.gitold'
Initialized empty Git repository in C:/Users/Chloe/Documents/GitHub/gottesacker/.git/
fatal: protocol '' is not supported
```

I am running these from Git Bash, in the repository that is having issues. 

Second question:
How do you know that it is the local directory that is corrupt? I'm hoping to get a sense of how to solve this kind of problem on my own so that I am not just blindly following instructions, but I'm having a hard time deciphering the error messages.

**Comment by steveward** [comment]
@cfandel the `fatal: protocol '' is not supported` error probably means a rogue space got placed somewhere in a command you were running. If you are running the commands individually you do not need to add the && at the end, and make sure you substitute the URL in the `git remote add` step to something like this:

`git remote add origin https://github.com/desktop/desktop`

>How do you know that it is the local directory that is corrupt? I'm hoping to get a sense of how to solve this kind of problem on my own so that I am not just blindly following instructions, but I'm having a hard time deciphering the error messages.

GitHub does a check on your repository when you attempt to push the repository up to GitHub. If any corruption is detected the push will be prevented to keep the remote repository in a healthy state. You also ran the `git fsck` command on your local repository, which is showing that there is corruption. I hope that helps explain things further. Let me know if I can answer any further questions.

**Comment by cfandel** [comment]
OK that's helpful - thanks!
I was able to fix the problem by running each command individually.
I also had initially put quotation marks around the URL (it's confusing in the original post that the placeholder is surrounded by brackets AND quote marks - makes it seem like the quotes are necessary).
Thank you so much for your patience and help!

PS: I think the original problem must have been due to a sudden power failure, because I did have one right before the problem showed up.
